pkg/app/test_command.go

```diff
@@ -2,53 +2,112 @@ package app
 
 import (
 	"fmt"
-	"github.com/SimonBaeumer/commander/pkg/output"
-	"github.com/SimonBaeumer/commander/pkg/runtime"
-	"github.com/SimonBaeumer/commander/pkg/suite"
 	"io/ioutil"
 	"log"
 	"os"
+	"sync"
+
+	"github.com/SimonBaeumer/commander/pkg/output"
+	"github.com/SimonBaeumer/commander/pkg/runtime"
+	"github.com/SimonBaeumer/commander/pkg/suite"
 )
 
 // TestCommand executes the test argument
 // file is the path to the configuration file
 // title ist the title of test which should be executed, if empty it will execute all tests
 // ctx holds the command flags
-func TestCommand(file string, title string, ctx AddCommandContext) error {
+func TestCommand(input string, title string, ctx AddCommandContext) error {
 	if ctx.Verbose == true {
 		log.SetOutput(os.Stdout)
 	}
 
-	if file == "" {
-		file = CommanderFile
+	if input == "" {
+		input = CommanderFile
 	}
-
-	fmt.Println("Starting test file " + file + "...")
-	fmt.Println("")
-	content, err := ioutil.ReadFile(file)
+	inputType, err := os.Stat(input)
 	if err != nil {
 		return fmt.Errorf("Error " + err.Error())
 	}
 
+	var results <-chan runtime.TestResult
+	if inputType.IsDir() {
+		fmt.Println("Starting test against directory: " + input + "...")
+		fmt.Println("")
+		results, err = testDir(input, title)
+	} else {
+		fmt.Println("Starting test file " + input + "...")
+		fmt.Println("")
+		results, err = testFile(input, title)
+	}
+
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	out := output.NewCliOutput(!ctx.NoColor, ctx.FileOrder)
+	if !out.Start(results) {
+		return fmt.Errorf("Test suite failed, use --verbose for more detailed output")
+	}
+
+	return nil
+}
+
+func testDir(directory string, title string) (<-chan runtime.TestResult, error) {
+	files, err := ioutil.ReadDir(directory)
+	if err != nil {
+		return nil, fmt.Errorf("Error " + err.Error())
+	}
+
+	results := make(chan runtime.TestResult)
```

Comment:
Using the `TestResult` type outside of the `runtime` package is irritating because it is intended to be just the result of the test suite. BUT the major problem with this is that the output package relies on the result channel and has too many responsiblities.

The output package does:
 - printing results immediately after receiving a result from the channel, see cli.go:47
 - aggregating results
 - print a summary  
 - measuring the duration

If we would take another approach - the `runtime` does not return a `channel` instead we add the output to the runtime struct and use it directly for the output.
The return value of the `runtime.Start` function could be a result which already contains the duration of the execution and some metadata about the whole test run - than the output package could easily print the summary based on the result of our runtime.

If we now want to add the directory feature we could aggregate all test results and get an overall overview of the complete run AND could provide insights into the execution of the each test file.

I hope this was understandable - basically the output package has to many responsibilities which leads to more complex solutions.